### PR TITLE
Add additional check for finished editing sign (fix #1063)

### LINKED ISSUE #1063: non-editable sign
**Describe the bug**
Unable to place text on lines 2-4 on signs using various versions of Bedrock

**To Reproduce**
On Android or Nintendo Switch (Xbox1, PS4 remain untested)
1. Place a sign
2. Place text on lines 2-4
3. Close GUI
4. Poof, no text on lines 2-4

**Expected behavior**

**Screenshots / Videos**
This warning appears whenever return/enter is pressed while editing a sign:
image

**Server Version**
image

**Geyser Version**
image

**Minecraft: Bedrock Edition Version**
Android, Nintendo Switch (confirmed not working)
Signs seems to be editable on Win10, iPad.

### PR DESCRIPTION
**Pull request by starcatmeow**
In #1063, some bedrock clients are unable to write the sign, throwing 'change non-editable sign'. It's probably caused by duplicated packets from the client which corrupts the original finishing check.

Though there's no direct field in the packet that indicates the user finishes editing, I'm only able to use some tricky way to determine when should we send the packet. In this PR, I delay the sign update packet until when player moving, doing action, editing another sign or disconnecting. Hope it helps and not causes more problem.

And if there's better(or less tricky) way to solve this, I'd love to hear from you.😄

### CODE REVIEW

**Comment by starcatmeow** [comment]
Ah, what if the user finishes editing then directly disconnect from server? I'm not sure.

File: core/src/main/java/org/geysermc/geyser/session/GeyserSession.java [hunk 1]
```diff
@@ -544,7 +554,14 @@ public GeyserSession(GeyserImpl geyser, BedrockServerSession bedrockServerSessio
         bedrockServerSession.addDisconnectHandler(disconnectReason -> {
             InetAddress address = bedrockServerSession.getRealAddress().getAddress();
             geyser.getLogger().info(GeyserLocale.getLocaleStringLog("geyser.network.disconnect", address, disconnectReason));
-
+            // Check if there is undetermined sign update packet
+            if (this.getLastSignUpdatePacket() != null) {
+                // We send the packet to the server
+                this.sendDownstreamPacket(this.getLastSignUpdatePacket());
+                // We set the sign text&packet cached in the session to null to indicate there is no work-in-progress sign
+                this.setLastSignMessage(null);
```
[the remark this answers is not in the document]
Comment (starcatmeow) [reply core/src/main/java/org/geysermc/geyser/session/GeyserSession.java:562]: Ah, what if the user finishes editing then directly disconnect from server? I don't know where to add the code for this case.

[the remark this answers is not in the document]
Comment (Camotoy) [reply core/src/main/java/org/geysermc/geyser/session/GeyserSession.java:562]: There should be GeyserSession#disconnect.

[the remark this answers is not in the document]
Comment (starcatmeow) [reply core/src/main/java/org/geysermc/geyser/session/GeyserSession.java]: Thanks, I moved it to the disconnect method.

File: core/src/main/java/org/geysermc/geyser/session/GeyserSession.java [hunk 2]
```diff
@@ -544,7 +554,6 @@ public GeyserSession(GeyserImpl geyser, BedrockServerSession bedrockServerSessio
         bedrockServerSession.addDisconnectHandler(disconnectReason -> {
             InetAddress address = bedrockServerSession.getRealAddress().getAddress();
             geyser.getLogger().info(GeyserLocale.getLocaleStringLog("geyser.network.disconnect", address, disconnectReason));
-
```
Comment (Camotoy) [inline core/src/main/java/org/geysermc/geyser/session/GeyserSession.java:547]: Re-add this line for a cleaner diff, please!

**Comment by Camotoy** [comment]
Two other thoughts:
- Do we need the previous sign check anymore?
- Maybe combine these checks with the book edit check? I can do this later outside of this PR.

**Comment by starcatmeow** [comment]
Thanks for your thoughts. I combined these checks with the book edit check. And I found out the previous sign case was covered by the update at inventory transaction packet, now I removed that check.

**Review by OskarWiedeweg — APPROVED** [review]

**Comment by Camotoy** [comment]
So I tested these changes, and they work. But I'm very conflicted if we should implement this. The current sign code, while it doesn't work for all instances, is much cleaner in how it's sent to the Java server. Unlike books, if the player gets hit while writing the sign, they're locked out of editing the sign (and likely very confused on what happened). We could add a config option, but I would hate a config option for a bug fix like this. But I also understand that editing a sign and seeing the behavior this PR fixes also sucks!
I made a proof-of-concept where we listened to a Bukkit sign change event and allowed the player to always edit the sign, which modifies server behavior and requires Bukkit, but leads to a bit friendlier user experience.
I'm not sure where to go from here. Advice wanted.

**Comment by Comstepr** [comment]
In my opinion, I support this PR because I think that this should be implemented -- signs are a crucial part of the game and I think it would be nice for Bedrock Edition players to be able to use them.

Do we have an estimate on when this will be pushed to master?

On another note, I'll be testing this on my production server soon. However, are there any caveats with this PR? Will this affect anything?

**Comment by Camotoy** [comment]
There is no estimate. The two downsides of this PR are that Bedrock players that get hit are locked out of the sign, and the sign changes will not apply until they move (once they are done). I'm working on, at the least, providing a hacky option for Paper servers to work around this.

**Comment by Camotoy** [comment]
The hacky option for Paper: https://github.com/Camotoy/GeyserHacks

**Comment by Comstepr** [comment]
Thanks @Camotoy , in that case do I need this PR or your GeyserHacks plugin?

**Comment by Camotoy** [comment]
One or the other.

**Comment by Comstepr** [comment]
Gotcha. Thanks!

**Comment by Camotoy** [comment]
I'm going to go ahead and close this issue as it seems that #1063 is resolved in 1.18.30 and 1.19 keeps this behavior, so it's intentional. We'll re-look at this PR if the behavior pops up again, but I think we're set for now. Thanks!

**Pull request closed without merging**